File: c++/advanced/1a_circular_array.cpp
//Find largest sum subarray (non-empty) in circular array: 

class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int currMax = 0, globalMax = INT_MIN;
        int currMin = 0, globalMin = INT_MAX;
        int total = 0;
        for(int n:nums){
            currMax = max(n, currMax+n);
            currMin = min(n, currMin+n);
            globalMax = max(globalMax, currMax);
            globalMin = min(globalMin, currMin);
            total+= n;
        }
        return globalMax>0 ? max(globalMax, total - globalMin) : globalMax ;
    }
};

File: js/advanced/1a_circular_array.js
//Find largest sum subarray (non-empty) in circular array: 

var maxSubarraySumCircular = function(nums) {
    let currMax = 0, globalMax = Number.MIN_SAFE_INTEGER
    let currMin = 0, globalMin = Number.MAX_SAFE_INTEGER
    let total = 0
    for(let n of nums){
        currMax = Math.max(currMax+n, n)
        currMin = Math.min(currMin+n, n)
        globalMax = Math.max(globalMax, currMax)
        globalMin = Math.min(globalMin, currMin)
        total+= n
    }
    return globalMax>0 ? Math.max(globalMax, total-globalMin) : globalMax
};